### PR TITLE
CLN/TYP: address TODOs, ignores

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/ops/array_ops.py [hunk 1]
```diff
@@ -91,11 +94,9 @@ def _masked_arith_op(x: np.ndarray, y, op):
     assert isinstance(x, np.ndarray), type(x)
     if isinstance(y, np.ndarray):
         dtype = find_common_type([x.dtype, y.dtype])
-        # error: Argument "dtype" to "empty" has incompatible type
-        # "Union[dtype, ExtensionDtype]"; expected "Union[dtype, None, type,
-        # _SupportsDtype, str, Tuple[Any, int], Tuple[Any, Union[int,
-        # Sequence[int]]], List[Any], _DtypeDict, Tuple[Any, Any]]"
-        result = np.empty(x.size, dtype=dtype)  # type: ignore[arg-type]
+        # x and y are both ndarrays -> common_dtype is np.dtype
```
Comment (simonjayhawkins) [inline pandas/core/ops/array_ops.py:97]: maybe we should overload `find_common_type`. could add a `# TODO`

File: pandas/core/reshape/reshape.py [hunk 2]
```diff
@@ -1046,9 +1046,7 @@ def get_empty_frame(data) -> DataFrame:
         fill_value: bool | float | int
         if is_integer_dtype(dtype):
             fill_value = 0
-        # error: Non-overlapping equality check (left operand type: "dtype[Any]", right
-        # operand type: "Type[bool]")
-        elif dtype == bool:  # type: ignore[comparison-overlap]
```
Comment (simonjayhawkins) [inline pandas/core/reshape/reshape.py:1051]: looks like a false positive (maybe from numpy types)

File: pandas/core/strings/accessor.py [hunk 3]
```diff
@@ -628,30 +628,22 @@ def cat(self, others=None, sep=None, na_rep=None, join="left"):
             # no NaNs - can just concatenate
             result = cat_safe(all_cols, sep)
 
+        out: Index | Series
```
Comment (simonjayhawkins) [inline pandas/core/strings/accessor.py:631]: can you also add this as the return type of the function signature.

File: pandas/core/strings/accessor.py [hunk 4]
```diff
@@ -628,30 +628,22 @@ def cat(self, others=None, sep=None, na_rep=None, join="left"):
...
         if isinstance(self._orig, ABCIndex):
             # add dtype for case that result is all-NA
 
-            # error: Incompatible types in assignment (expression has type
-            # "Index", variable has type "ndarray")
-            result = Index(  # type: ignore[assignment]
-                result, dtype=object, name=self._orig.name
-            )
+            out = Index(result, dtype=object, name=self._orig.name)
```
Comment (simonjayhawkins) [inline pandas/core/strings/accessor.py:635]: maybe could just return here and no need for the variable type annotation.

File: pandas/io/stata.py [hunk 5]
```diff
@@ -934,25 +934,15 @@ def __eq__(self, other: Any) -> bool:
 
     @classmethod
     def get_base_missing_value(cls, dtype: np.dtype) -> int | float:
-        # error: Non-overlapping equality check (left operand type: "dtype[Any]", right
-        # operand type: "Type[signedinteger[Any]]")
-        if dtype == np.int8:  # type: ignore[comparison-overlap]
+        if dtype.type == np.int8:
```
Comment (simonjayhawkins) [inline pandas/io/stata.py:937]: does this still work with `is`?

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @jbrockmendel generally LGTM.

File: pandas/core/internals/array_manager.py [hunk 6]
```diff
@@ -1109,8 +1109,8 @@ def as_array(
         arr : ndarray
         """
         if len(self.arrays) == 0:
-            arr = np.empty(self.shape, dtype=float)
-            return arr.transpose() if transpose else arr
+            farr = np.empty(self.shape, dtype=float)
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/internals/array_manager.py:1112]: > while on the subject, I've never guessed or seen anything to give a clue as to what tipo is?

i think its DtypeObj; no idea on why the name was chosen

File: pandas/core/internals/base.py [hunk 7]
```diff
@@ -157,7 +157,7 @@ class SingleDataManager(DataManager):
 
     @final
     @property
-    def array(self):
+    def array(self) -> ArrayLike:
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/internals/base.py:160]: will do

File: pandas/io/stata.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/io/stata.py:937]: i think so, will check

File: pandas/core/reshape/reshape.py (hunk 2 above)
Comment (twoertwein) [reply pandas/core/reshape/reshape.py:1051]: `np.bool_` does not inherit from `bool`: https://numpy.org/doc/stable/reference/arrays.scalars.html

Comment (simonjayhawkins) [reply pandas/core/reshape/reshape.py:1051]: The error message is `Non-overlapping equality check...`. The numpy types should allow a check with a type. either `bool`, `np.bool_` and `str`, `object` or any other type. Doing the equality check does not raise a type error, but should always return bool.

```
import numpy as np

dtype = np.dtype("bool")
print(dtype)
print(dtype == bool)
print(dtype == np.bool_)
print(dtype == "bool")
print(dtype == object)
```
gives
```
bool
True
True
True
False
```

In the numpy type annotations for 1.21.3,  dtype.\_\_eq__ is not defined, but looks like this is fixed on master. https://github.com/numpy/numpy/pull/20184

https://github.com/numpy/numpy/blob/8eb6555711888f0ab5594a2bd82d230ba33a3d7c/numpy/__init__.pyi#L876-L880

Comment (jbrockmendel) [reply pandas/core/reshape/reshape.py:1051]: so what's the suggested action here?

Comment (simonjayhawkins) [reply pandas/core/reshape/reshape.py:1051]: what you've done is fine. but we could also leave these ignores for now. 

Once we are checking with the latest numpy types we could maybe revert all the changes that have been made to-date to get round this. wdyt?

Comment (jbrockmendel) [reply pandas/core/reshape/reshape.py:1051]: sure

**Review by simonjayhawkins — APPROVED** [review]

**Comment by simonjayhawkins** [comment]
Thanks @jbrockmendel

**Pull request merged**